File: dashboard/lib/domain/slope-cells.ts
// slope_cells.json 로더 + 픽셀<->미터 변환 (브리프 D1/D3, 스펙 §7.3·§3.5)
//
// 이 파일은 엔진 산출물 slope_cells.json(engine/flatness/outputs/slope_cells.py)의
// 기하 정보(SlopeCell 13필드)만 다룬다. 그 파일에는 판정 결과(grade·dev_pct·
// correction_mm·dir_err_deg)가 없다 - 재판정 때 다시 계산되는 파생값이고 같이
// 담으면 CSV와 이중 진실이 되기 때문이다(브리프 D1). 그래서 이 로더도 등급을
// 계산하거나 다루지 않는다: 화면 조립(Task 5)이 별도로 얻은 grade를 여기서 낸
// 기하 정보와 짝지어 slope-heatmap.ts 렌더러에 넘긴다.
//
// BOM 함정 없음: slope_cells.json은 utf-8이다(CSV만 utf-8-sig로 저장된다).
// CSV는 재판정 입력으로 쓰면 반올림 때문에 판정이 달라지는 것이 실측으로
// 확인됐다(브리프 D1) - 그래서 이 파일은 CSV를 전혀 파싱하지 않는다.
import { dataUrl } from './paths';
import type { SlopeCellRow, SlopeCellsFile, SlopeStats } from './types';

function isFiniteNumber(v: unknown): v is number {
  return typeof v === 'number' && Number.isFinite(v);
}

function isNumberOrNull(v: unknown): v is number | null {
  return v === null || isFiniteNumber(v);
}

/** 코드리뷰 M4: 최상위 키만 보고 cells 배열 원소는 검증하지 않으면, width_m
 * 누락이나 좌표가 문자열인 행도 그대로 통과해 이후 산술(slope-heatmap.ts의
 * worldToPx 등)에서 문자열이 조용히 숫자로 강제변환되거나 NaN이 퍼진다 - 이
 * 저장소가 가장 경계하는 조용한 실패다. SlopeCell 13필드를 전부 타입 수준까지
 * 확인한다. */
function isValidSlopeCellRow(row: unknown): row is SlopeCellRow {
  if (!row || typeof row !== 'object') return false;
  const r = row as Record<string, unknown>;
  return isFiniteNumber(r.cx) && isFiniteNumber(r.cy)
    && isFiniteNumber(r.center_x) && isFiniteNumber(r.center_y)
    && isFiniteNumber(r.n_subcells)
    && isNumberOrNull(r.slope_pct) && isNumberOrNull(r.downhill_rad)
    && isNumberOrNull(r.rmse_m) && isNumberOrNull(r.se_pct)
    && isFiniteNumber(r.width_m) && isFiniteNumber(r.height_m)
    && typeof r.ok === 'boolean'
    && (r.zone_id === null || isFiniteNumber(r.zone_id));
}

/** slope_cells.json으로 파싱된 원시 객체가 실제로 그 형태인지 내용으로 판별한다.
 *
 * stats.json의 isSlopeStats(stats.ts)와 같은 관례 - 별도 스키마 라이브러리 없이
 * 이 저장소는 최상위 판별 키를 직접 확인하는 가벼운 타입가드를 쓴다. 다만
 * cells 배열은 원소 하나하나가 SlopeCellRow 형태인지까지 확인한다(코드리뷰 M4 -
 * 배열 존재 여부만 보면 원소가 깨져도 통과한다).
 */
export function isSlopeCellsFile(raw: unknown): raw is SlopeCellsFile {
  if (!raw || typeof raw !== 'object') return false;
  const r = raw as Record<string, unknown>;
  return typeof r.schema_version === 'number'
    && typeof r.engine_version === 'string'
    && typeof r.cell_m === 'number'
    && typeof r.subcell_m === 'number'
    && Array.isArray(r.cells)
    && r.cells.every(isValidSlopeCellRow);
}

/** stats.artifacts.cells_json -> /api/data 경로.
 *
 * 경로 함정(브리프 D3/Task 4): stats.artifacts 값은 이미 버킷-상대 전체 경로다
 * (worker/flatworker/slope.py가 artifacts/{id}/...로 정규화한다). 평활도처럼
 * artifactUrl(artifacts_dir, name)을 쓰면 artifacts/{id}/artifacts/{id}/...로
 * 중복돼 404가 난다 - 그래서 dataUrl()에 값을 그대로 넘긴다.
 *
 * cells_json이 없으면(브리프 D7: 단계 C까지 만들어진 분석) null - 호출부가 이
 * 상태를 명시적으로 다뤄야 한다(히트맵 대신 slope_map.png, 클릭 비활성화).
 */
export function slopeCellsJsonUrl(artifacts: SlopeStats['artifacts']): string | null {
  if (!artifacts.cells_json) return null;
  return dataUrl(artifacts.cells_json);
}

/** 셀들이 차지하는 실좌표(미터) 경계. gridGeometry(lib/viz/heatmap.ts)와 달리
 * 격자 인덱스가 아니라 월드 좌표를 낸다 - 배수구 클릭이 연속 미터 좌표이므로
 * (스펙 §3.5) 격자 인덱스만으로는 클릭 위치를 복원할 수 없다. */
export interface SlopeWorldBounds { xMin: number; xMax: number; yMin: number; yMax: number; }

export function slopeWorldBounds(cells: SlopeCellRow[]): SlopeWorldBounds | null {
  if (cells.length === 0) return null;
  let xMin = Infinity, xMax = -Infinity, yMin = Infinity, yMax = -Infinity;
  for (const c of cells) {
    const hw = c.width_m / 2;
    const hh = c.height_m / 2;
    if (c.center_x - hw < xMin) xMin = c.center_x - hw;
    if (c.center_x + hw > xMax) xMax = c.center_x + hw;
    if (c.center_y - hh < yMin) yMin = c.center_y - hh;
    if (c.center_y + hh > yMax) yMax = c.center_y + hh;
  }
  return { xMin, xMax, yMin, yMax };
}

/** 월드(미터) <-> Canvas 픽셀 변환 계수. scale은 미터당 픽셀 수. */
export interface SlopeTransform { scale: number; xMin: number; yMax: number; }

/** 경계를 maxW x maxH 픽셀 안에 맞추는 스케일을 구한다(가로세로 비율 유지). */
export function slopeTransformFor(bounds: SlopeWorldBounds, maxW: number, maxH: number): SlopeTransform {
  const w = Math.max(bounds.xMax - bounds.xMin, 1e-6);
  const h = Math.max(bounds.yMax - bounds.yMin, 1e-6);
  const scale = Math.max(Math.min(maxW / w, maxH / h), 1e-6);
  return { scale, xMin: bounds.xMin, yMax: bounds.yMax };
}

/** 월드 좌표(미터, y 위로 증가) -> Canvas 픽셀(y 아래로 증가).
 *
 * ★ y축 함정(브리프 D3): 엔진 downhill_rad·center_y는 matplotlib 데이터 좌표계
 * (y 위로 증가) 기준이다. Canvas는 y가 아래로 증가하므로 여기서 반드시 뒤집는다.
 * lib/viz/heatmap.ts의 cellRect가 격자 행(iy)만 뒤집는 선례를 실좌표(미터)에도
 * 그대로 적용한 것이다. */
export function worldToPx(t: SlopeTransform, x: number, y: number): { px: number; py: number } {
  return { px: (x - t.xMin) * t.scale, py: (t.yMax - y) * t.scale };
}

/** Canvas 픽셀 -> 월드 좌표(미터). worldToPx의 역함수.
 *
 * 배수구 클릭(Task 5)이 이 함수를 쓴다. 스펙 §3.5의 배수구 좌표는 연속 미터
 * 좌표({"x":3.2,"y":5.1})이므로 셀 중심으로 스냅하지 않고 이 값을 그대로 잡 payload에
 * 싣는다(2m 해상도로 스냅하면 배수구 위치가 지나치게 굵어진다). */
export function pxToWorld(t: SlopeTransform, px: number, py: number): { x: number; y: number } {
  return { x: t.xMin + px / t.scale, y: t.yMax - py / t.scale };
}

File: engine/flatness/outputs/slope_cells.py
"""slope_cells.json 왕복 직렬화 — 재판정(§7.3)과 화면 히트맵 재구성이 함께 소비하는
무손실 셀 산출물.

CSV(slope_cells.csv)가 아니라 이 파일을 재판정 입력으로 쓰는 이유는 브리프 D1의
실측 때문이다:
  - CSV에는 width_m/height_m 열이 없어 명목 cell_m으로 복원하면 correction_mm이
    가장자리 셀에서 정확히 2배로 나온다(실측 0.87 -> 1.74mm)
  - 조각 셀의 판정불가 사유가 뒤바뀐다("격자 가장자리 조각 셀" -> "유효 서브셀 부족")
  - CSV의 반올림(소수 3자리·각도 1자리)이 판정 경계에 걸터앉은 바닥의 등급을
    뒤집는다(20개 중 11개, 실측)

그래서 SlopeCell 12필드(+zone_id) 전부를 반올림 없이 그대로 담는다. 판정 결과
(grade·dev_pct·correction_mm·dir_err_deg·reason)는 담지 않는다 - 재판정 때
grade_slope_cells가 다시 계산하는 파생값이고, 같이 담으면 CSV와 이중 진실이 된다.

**cell_m/subcell_m도 함께 담는다(Task 1 리뷰 I1).** grade_slope_cells의 조각 셀
분기(`width_m < cell_m/2`)가 cell_m에 의존하는데, 재판정 호출부가 원 분석과 다른
cell_m을 쓰면 CSV 근사 복원과 똑같은 결함(D1이 배제하려던 바로 그것)이 JSON
경로에서 재현된다. cell_m을 파일에 함께 실어 두면 재판정이 이 값을 그대로
읽어 쓸 수 있다 - "기본값을 조용히 쓰다 잊는" 경로를 근본적으로 없앤다.

선례: 평활도가 cells.json을 별도로 내고(outputs/stats.py) 대시보드가 fetch한다.
같은 파일을 재판정 잡과 화면이 함께 소비하면 히트맵 재구성과 재판정 입력이
한 파일로 해결된다.
"""
import json
import math
from dataclasses import fields

from flatness.core.slope import SlopeCell

# 파일 구조(필드 구성)가 바뀔 때만 올린다. 2 -> cell_m/subcell_m 최상위 키 추가
# (Task 1 리뷰 I1). 이 워크트리 안에서만 존재하던 산출물이라 과거 버전과의
# 하위 호환은 고려하지 않는다(실제 배포 전이므로 기존 파일이 없다).
SCHEMA_VERSION = 2

# not ok인 셀에서 nan이 나오는 필드. compute_slope_cells가 판정 불가 셀에
# float("nan")을 채우는 바로 그 네 필드다(slope.py 참고).
_NAN_FIELDS = ("slope_pct", "downhill_rad", "rmse_m", "se_pct")

_FIELD_NAMES = tuple(f.name for f in fields(SlopeCell))


def _cell_to_row(c):
    """SlopeCell -> JSON에 쓸 dict. nan은 null로 바꾼다.

    이 저장소 관례가 json.dump(allow_nan=False)다(pipeline.py) - RFC 8259
    표준 JSON에는 NaN 토큰이 없어 브라우저 JSON.parse·Postgres jsonb가 조용히
    거부하기 때문이다. 여기서도 같은 관례를 따른다.
    """
    row = {}
    for name in _FIELD_NAMES:
        v = getattr(c, name)
        if name in _NAN_FIELDS and isinstance(v, float) and math.isnan(v):
            row[name] = None
        else:
            row[name] = v
    return row


def _row_to_cell(row):
    """JSON dict -> SlopeCell. null이던 nan 필드를 float('nan')으로 되돌린다."""
    kwargs = dict(row)
    for name in _NAN_FIELDS:
        if kwargs.get(name) is None:
            kwargs[name] = float("nan")
    return SlopeCell(**kwargs)


def dump_slope_cells(cells, path, engine_version, cell_m, subcell_m):
    """SlopeCell 리스트 + 원 분석의 격자 파라미터를 무손실 JSON으로 쓴다.

    cell_m/subcell_m은 기본값을 두지 않고 필수 인자로 받는다 - analyze_slope가
    이미 알고 있는 값을 그대로 넘기게 강제해, 나중에 이 호출부를 고치다 값을
    빠뜨리는 실수를 호출 시점에 바로 TypeError로 드러내기 위해서다(Task 1
    리뷰 I1: 기본값이 있으면 재판정이 조용히 다른 cell_m으로 돈다).

    engine_version은 compute_slope_cells를 낸 엔진 빌드를 기록한다. 재판정
    (judge_slope_cells)이 이 값을 검사하지는 않는다 - 판정 로직(grade_slope_cells)
    자체가 버전 사이에 안정적이라는 전제 위에서, SlopeCell 12필드의 의미가
    그대로인 한 엔진 버전이 달라도 무방하다고 본다.

    한계(Task 1 리뷰 I3): 이 전제가 깨지면(즉 grade_slope_cells의 판정 로직
    자체가 엔진 버전 사이에 바뀌면) 재판정이 원 분석 발행 시점과 다른 결과를
    조용히 낼 수 있다. 지금은 그 로직 변경을 감지·차단하는 장치가 없다 - 버전
    간 판정 로직 마이그레이션 정책은 이 태스크 범위 밖의 후속 과제다. 이 한계는
    load_slope_cells가 돌려주는 meta에도 engine_version이 실려 있어(호출부가
    필요하면) 확인할 수 있게만 해 둔다.
    """
    payload = {
        "schema_version": SCHEMA_VERSION,
        "engine_version": engine_version,
        "cell_m": cell_m,
        "subcell_m": subcell_m,
        "cells": [_cell_to_row(c) for c in cells],
    }
    with open(path, "w", encoding="utf-8") as f:
        # allow_nan=False: _cell_to_row에서 이미 nan을 null로 치환했다. 그런데도
        # nan이 남아 있다면(치환 누락 등 버그) 조용히 NaN 토큰을 내보내느니 여기서
        # 바로 예외로 터뜨리는 편이 낫다(pipeline.py와 동일한 관례).
        json.dump(payload, f, ensure_ascii=False, indent=2, allow_nan=False)
    return path


def load_slope_cells(path):
    """slope_cells.json -> (cells, meta). dump_slope_cells의 역함수다.

    meta = {"schema_version", "engine_version", "cell_m", "subcell_m"}.
    cells만 필요해서 meta를 버리면(list[SlopeCell]만 받던 이전 인터페이스)
    호출부가 cell_m을 다시 기본값으로 채워 넣기 쉽다 - 그게 바로 Task 1 리뷰
    I1이 지적한 결함이다. 그래서 항상 튜플로 묶어 반환해, 재판정 호출부가
    `cells, meta = load_slope_cells(path)` 다음 `judge_slope_cells(cells, ...,
    meta["cell_m"], subcell_m=meta["subcell_m"])`처럼 원 분석의 격자 파라미터를
    그대로 이어 쓰도록 유도한다.

    engine_version의 한계(버전 간 grade_slope_cells 변경은 감지하지 못함)는
    dump_slope_cells 독스트링 참고 - 여기서 meta로 노출하는 것으로 그 판단을
    호출부(judge_slope_cells 포함)에서도 볼 수 있게 한다.
    """
    with open(path, encoding="utf-8") as f:
        payload = json.load(f)
    cells = [_row_to_cell(row) for row in payload["cells"]]
    meta = {
        "schema_version": payload["schema_version"],
        "engine_version": payload["engine_version"],
        "cell_m": payload["cell_m"],
        "subcell_m": payload["subcell_m"],
    }
    return cells, meta
